cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/update/StateVariablesAdder.java

```diff
@@ -7,86 +7,205 @@
+            return;
+        }
+        // Clear the previous SV data in CGMES model
+        // and fill it with the Network current state values
+        cgmes.clear(CgmesSubset.STATE_VARIABLES);
+
+        if (cimVersion != 14) {
+            addModelDescriptionToCgmes();
+            addTopologicalIslandsToCgmes();
+        }
+
+        addVoltagesForTopologicalNodes();
+
+        addVoltagesForBoundaryNodes();
+
+        addPowerFlowToCgmes();
 
+        addShuntCompensatorSectionsToCgmes();
+
+        addTapStepToCgmes();
+
+        addStatusToCgmes();
+    }
+
+    private void addVoltagesForTopologicalNodes() {
         PropertyBags voltages = new PropertyBags();
-        for (Bus b : n.getBusBreakerView().getBuses()) {
+        // add voltages for TpNodes existing in the Model
+        for (PropertyBag tn : cgmes.topologicalNodes()) {
+            Bus b = network.getBusBreakerView().getBus(tn.getId(CgmesNames.TOPOLOGICAL_NODE));
             PropertyBag p = new PropertyBag(SV_VOLTAGE_PROPERTIES);
-            p.put(CgmesNames.ANGLE, fs(b.getAngle()));
-            p.put(CgmesNames.VOLTAGE, fs(b.getV()));
-            p.put("TopologicalNode", topologicalNodeFromBusId(b.getId()));
-            voltages.add(p);
+            if (b != null) {
+                p.put(CgmesNames.ANGLE, fs(b.getAngle()));
+                p.put(CgmesNames.VOLTAGE, fs(b.getV()));
+                p.put(CgmesNames.TOPOLOGICAL_NODE, tn.getId(CgmesNames.TOPOLOGICAL_NODE));
+                voltages.add(p);
+            }
         }
-        cgmes.add(CgmesSubset.STATE_VARIABLES, "SvVoltage", voltages);
+        cgmes.add(originalSVcontext, "SvVoltage", voltages);
+    }
 
+    private void addVoltagesForBoundaryNodes() {
+        PropertyBags voltages = new PropertyBags();
+        // add voltages for TpNodes existing in the Model's boundaries
+        boundaryNodesFromDanglingLines.entrySet().forEach(entry -> {
+            PropertyBag p = new PropertyBag(SV_VOLTAGE_PROPERTIES);
+            DanglingLine dl = network.getDanglingLine(entry.getKey());
+            Bus b = dl.getTerminal().getBusBreakerView().getBus();
+            if (b != null) {
+                // calculate complex voltage value: abs for VOLTAGE, degrees for ANGLE
+                Complex v2 = complexVoltage(dl.getR(), dl.getX(), dl.getG(), dl.getB(), b.getV(), b.getAngle(),
+                    dl.getP0(), dl.getQ0());
+                p.put(CgmesNames.ANGLE, fs(Math.toDegrees(v2.getArgument())));
+                p.put(CgmesNames.VOLTAGE, fs(v2.abs()));
+                p.put(CgmesNames.TOPOLOGICAL_NODE, entry.getValue());
+            } else {
+                p.put(CgmesNames.ANGLE, fs(0.0));
+                p.put(CgmesNames.VOLTAGE, fs(0.0));
+                p.put(CgmesNames.TOPOLOGICAL_NODE, entry.getValue());
+            }
+            voltages.add(p);
+        });
+        cgmes.add(originalSVcontext, "SvVoltage", voltages);
+    }
+
+    private void addPowerFlowToCgmes() {
```

Comment:
To lower this method's complexity, create an intermediary method:
```java
private <I extends Injection> void addInjectionPowerFlowToCgmes(PropertyBags powerFlows, Iterable<I> injectionStream) {
        injectionStream.forEach(i -> {
            PropertyBag p = createPowerFlowProperties(i.getTerminal());
            if (p != null) {
                powerFlows.add(p);
            } else if (i instanceof Load) {
                // FIXME CGMES SvInjection objects created as loads
                LOG.error("No SvPowerFlow created for load {}", i.getId());
            }
        });
    }
```

You can then call:
```java
PropertyBags powerFlows = new PropertyBags();
addInjectionPowerFlowToCgmes(powerFlows, network.getLoads());
addInjectionPowerFlowToCgmes(powerFlows, network.getGenerators());
addInjectionPowerFlowToCgmes(powerFlows, network.getShuntCompensators());
addInjectionPowerFlowToCgmes(powerFlows, network.getStaticVarCompensators());
addInjectionPowerFlowToCgmes(powerFlows, network.getBatteries());
```

Reply:
Done.
Thank you Miora, that's really good, I'll try to use this pattern in future.